src/Tools/PrepareTests/TestDiscovery.cs

```diff
@@ -0,0 +1,127 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Diagnostics.Contracts;
+using System.IO;
+using System.Linq;
+using System.Text.Json;
+using Microsoft.TestPlatform.VsTestConsole.TranslationLayer;
+using Microsoft.VisualStudio.TestPlatform.ObjectModel;
+using Microsoft.VisualStudio.TestPlatform.ObjectModel.Client;
+using Microsoft.VisualStudio.TestPlatform.ObjectModel.Logging;
+
+namespace PrepareTests;
+internal class TestDiscovery
+{
+    public static void RunDiscovery(string sourceDirectory, string dotnetPath, bool isUnix)
+    {
+        var binDirectory = Path.Combine(sourceDirectory, "artifacts", "bin");
+        var assemblies = GetAssemblies(binDirectory, isUnix);
+
+        Console.WriteLine($"Found {assemblies.Count} test assemblies");
+
+        var vsTestConsole = Directory.EnumerateFiles(Path.Combine(Path.GetDirectoryName(dotnetPath)!, "sdk"), "vstest.console.dll", SearchOption.AllDirectories).Last();
```

Comment:
In case there are multiple sdk's in the directory.  This won't happen on CI machines but can happen locally.  In that case we just pick the latest (last) version.  It is a bit of an edge case where there is an sdk update in roslyn but the old one still exists in .dotnet

I realize now though that the order is dependent on file system impl, so updating this to explicitly order.